File: Sources/structly/05_binary_tree/TreeSum.swift
//
//  TreeSum.swift
//  MyCLI
//
//  Created by new on 3/24/26.
//


func treeSum(_ node: TreeNode<Int>?) -> Int {
    let q = Queue<TreeNode<Int>>()
    if let node {
        q.enqueue(node)
    }

    var res = 0
    while q.peek() != nil, let cnode = q.dequeue() {
        res += cnode.val
        if let left = cnode.left {
            q.enqueue(left )
        }
        if let right = cnode.right {
            q.enqueue(right)
        }
    }
    
    return res
}

func treeSum1(_ node: TreeNode<Int>?) -> Int {
    guard let node else {
        return 0
    }

    return node.val
        + (node.left.map(treeSum1(_:)) ?? 0)
        + (node.right.map(treeSum1(_:)) ?? 0)
}

File: Sources/structyC/05_binary_tree/tree_sum.cpp
//
//  tree_sum.cpp
//  MyCLI
//
//  Created by new on 3/24/26.
//

class Node {
public:
    int val;
    Node* left;
    Node* right;

    Node(int initialVal) {
        val = initialVal;
        left = nullptr;
        right = nullptr;
    }
};

int treeSum1(Node* root) {
    if (root == nullptr) {
        return 0;
    }

    return root->val
        + (root->left != nullptr ? treeSum1(root->left) : 0)
        + (root->right != nullptr ? treeSum1(root->right) : 0);
}
